src/shims/windows/sync.rs

```diff
@@ -130,4 +224,198 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
+                active_thread,
+                timeout,
+                Box::new(move |this| {
+                    this.reacquire_cond_lock(active_thread, lock_id, shared)?;
+
+                    this.condvar_remove_waiter(condvar_id, active_thread);
+
+                    let error_timeout = this.eval_windows("c", "ERROR_TIMEOUT")?;
+                    this.set_last_error(error_timeout)?;
+                    this.write_scalar(Scalar::from_i32(0), &dest)?;
+                    Ok(())
+                }),
+            );
+        }
+
+        this.write_scalar(Scalar::from_i32(1), dest)?;
+
+        Ok(())
+    }
+
+    fn WakeConditionVariable(&mut self, condvar_op: &OpTy<'tcx, Tag>) -> InterpResult<'tcx> {
+        let this = self.eval_context_mut();
+        let condvar_id = this.condvar_get_or_create_id(condvar_op)?;
+
+        if let Some((thread, lock, shared)) = this.condvar_signal(condvar_id) {
+            this.reacquire_cond_lock(thread, RwLockId::from_u32(lock), shared)?;
+            this.unregister_timeout_callback_if_exists(thread);
+        }
+
+        Ok(())
+    }
+
+    fn WakeAllConditionVariable(&mut self, condvar_op: &OpTy<'tcx, Tag>) -> InterpResult<'tcx> {
+        let this = self.eval_context_mut();
+        let condvar_id = this.condvar_get_or_create_id(condvar_op)?;
+
+        while let Some((thread, lock, shared)) = this.condvar_signal(condvar_id) {
+            this.reacquire_cond_lock(thread, RwLockId::from_u32(lock), shared)?;
+            this.unregister_timeout_callback_if_exists(thread);
+        }
+
+        Ok(())
+    }
+
+    fn WaitOnAddress(
+        &mut self,
+        ptr_op: &OpTy<'tcx, Tag>,
+        compare_op: &OpTy<'tcx, Tag>,
+        size_op: &OpTy<'tcx, Tag>,
+        timeout_op: &OpTy<'tcx, Tag>,
+        dest: &PlaceTy<'tcx, Tag>,
+    ) -> InterpResult<'tcx> {
+        let this = self.eval_context_mut();
+
+        let thread = this.get_active_thread();
+
+        let ptr = this.read_pointer(ptr_op)?;
+
+        let addr = ptr.addr().bytes();
+
+        let compare = this.deref_operand(compare_op)?;
+        let size = this.read_scalar(size_op)?.to_machine_usize(this)?;
+
+        let timeout_ms = this.read_scalar(timeout_op)?.to_u32()?;
+
+        let timeout_time = if timeout_ms == this.eval_windows("c", "INFINITE")?.to_u32()? {
+            None
+        } else {
+            this.check_no_isolation("`WaitOnAddress` with non-infinite timeout")?;
+
+            let duration = Duration::from_millis(timeout_ms as u64);
+
+            Some(Time::Monotonic(Instant::now().checked_add(duration).unwrap()))
+        };
+
+        let Some(layout) = this.wait_on_address_layout(size) else {
+            let invalid_param = this.eval_windows("c", "ERROR_INVALID_PARAMETER")?;
+            this.set_last_error(invalid_param)?;
+            this.write_scalar(Scalar::from_i32(0), dest)?;
+            return Ok(())
+        };
+
+        let futex_val = this.read_futex_val(ptr_op, layout)?;
```

Comment:
I think the same thing in https://github.com/rust-lang/miri/pull/2228 is required here: an SC fence before atomically reading the futex in wait and an SC fence somewhere in wake

Reply:
Are the semantics sufficiently similar that we can put more of this into shared code (like we do with mutex, condvar, rwlock)?